### PR TITLE
52 client integrations

### LINKED ISSUE #52: Integrate StoryRecordingClient with Admin/Metadata operations client
This ticket depends on Issue #33  and Issue #30 

The task is to integrate client RPC communication with the ChronoVisor process and client RPC communication with the ChronoKeeper processes under the same chronolog client   interface. The client should maintain  its connection state as well the state of  all the chronicles/stories it has acquired.

### PR DESCRIPTION
**Pull request by ibrodkin**
The changes in this pull request are the continuation of the changes made in #33 

At the top level the changes include:
1.  Updated and thread safe client facing API : chronoog::Client
2. actual client lib logic implementation is moved to chronolog::ChronologClientImpl
3. mutex protected state management is added to the Client that synchronizes the client  connect/disconnect/acquire/release/create/destroy calls for chronicles & stories from multiple threads 
4. AcquireStory API has templated StoryHandle added , implemented for writing case with StoryWritingHandle
5. StorytellerClient  implements  Keeper state handling logic , templated keeper selection logic, event timestamping and writing to the keeper processes . 
6. KeeperRecordingClients use the same thallium engine as the RPC Visor Client passed through the StorytellerClient to the KeeperRecordingclients that the later creates and owns
7. RPCClient & RPCVisor remote procedure calls on both sides are revised to have const correctness for the used parameters and the unused parameters removed throughout
8. Also made changes to the test  programs listed under test/integration/* to make sure they all still work with the changes chronolog client

### CODE REVIEW

File: Client/src/ChronologClientImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                             std::string const& client_account,
+                             int &flags)
+{
+    std::lock_guard<std::mutex> lock_client(chronologClientMutex);
+    // if already connected return success
+    // if disconencting return failure....
+    if((clientState != UNKNOWN) && (clientState != SHUTTING_DOWN))
+    {  return CL_SUCCESS;  }  
+    else if (clientState == SHUTTING_DOWN) 
+    {  return CL_ERR_INVALID_ARG; }
```
Comment (fkengun) [inline Client/src/ChronologClientImpl.cpp]: invalid_arg does not look intuitive. we can have a new err code for this.

File: Client/src/ChronologClientImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
+    {  return CL_ERR_NO_CONNECTION; }
+
+  //TODO : release all the acquired stories before asking to disconnect...
+
+    int flags=1;
+    auto return_code = rpcClient_->Disconnect(clientAccount , flags);
+    if(return_code == CL_SUCCESS)
+    {    
+	clientState = SHUTTING_DOWN;
```
Comment (fkengun) [inline Client/src/ChronologClientImpl.cpp:76]: does Disconnect leave something in the background before it returns? shutting_down is a bit ambiguous to me. To me, I think we should block Disconnect until everything is cleaned up for this particular connection. I would suggest something like DISCONNECTED or DOWN as the name after Disconnect.

File: Client/src/rpcVisorClient.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+
+    // constructor is private to make sure thalium rpc objects are created on the heap, not stack
+    RpcVisorClient( tl::engine & tl_engine, std::string const& service_addr, uint16_t provider_id)
+	    : service_addr(service_addr), service_provider_id(provider_id)
+	    , service_ph(tl_engine.lookup( service_addr),provider_id)
+	{
+	 std::cout<<" RpcVisorClient created for Visor Service at {"<<service_addr<<"} provider_id {"<<service_provider_id<<"}"<<std::endl;
+   	 visor_connect =tl_engine.define("Connect");
+   	 visor_disconnect =tl_engine.define("Disconnect");
```
Comment (fkengun) [inline Client/src/rpcVisorClient.h:114]: I am assuming tl::remote_procedures for metadata APIs will be added later.

File: Client/src/StorytellerClient.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+return ++atomic_index;
+}
+////////////////
+
+
+int chronolog::StorytellerClient::addKeeperRecordingClient( chronolog::KeeperIdCard const& keeper_id_card)
+{
+	   std::lock_guard<std::mutex> lock(recordingClientMapMutex);
+
+           std::string service_na_string("ofi+sockets://");
```
Comment (fkengun) [inline Client/src/StorytellerClient.cpp]: I am assuming protocol string from confmanager will be used later here.

**Review by fkengun — COMMENTED** [review]
Comments are left. no major changes required. 
Pls format the codes. I find myself difficult to read unaligned code when it gets long. Even simple `gg=G` in vim is a help to me.

File: Client/src/rpcVisorClient.h (hunk 3 above)
Comment (ibrodkin) [reply Client/src/rpcVisorClient.h:114]: rpcVisorClient.h  is intended to replace RPCClient.h and will inherit metadata APIs in the next iteration . it's not yet used anywhere in the code , RPCClient.h is still used used in this iteration

File: Client/src/StorytellerClient.cpp (hunk 4 above)
Comment (ibrodkin) [reply Client/src/StorytellerClient.cpp]: Yes, the protocol string will have to be passed in, not hardcoded . though I'm not sure we can assume that all the nodes will have the same fabrics available.

Comment (fkengun) [reply Client/src/StorytellerClient.cpp]: for the same reason, I split the communication protocol configurations for those three edges of the triangle into separate sections in the JSON conf file. I think tcp/socket will be supported anyway and verbs support depends on the available hardware. So during connect, the client library should try verbs first. If it fails, go to tcp/socket. Of course, we can leave it for later.

File: Client/src/ChronologClientImpl.cpp (hunk 2 above)
Comment (ibrodkin) [reply Client/src/ChronologClientImpl.cpp:76]: SHUTTING_DOWN is meant to be a final state , not transitional.  no new requests for data would be allowed if state = SHUTTING_DOWN

**Review by fkengun — APPROVED** [review]

**Pull request merged**